### ISSUE TITLE
making shader local to scene doesn't make it local to scene

### ISSUE THREAD

**Issue by lukostello** [godotengine/godot-proposals#2776]
### Describe the project you are working on
I am making a sokoban-like with twisty puzzle elements

### Describe the problem or limitation you are having in your project
When I check local to scene on the shader I expect the shader to be able to pass in unique parameters in each scene it is used in. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
when you check local to scene on the shader it should allow you to pass in unique parameters in each scene. Otherwise is it really doing anything at all? Is it making a copy of the shader code per scene? How would one even use that? There has got to be a way to just have the parameters being passed in be local to scene.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
I think it certainly will be used more often than whatever making shaders local to scene is doing right now. There is kind of one work around that I have been using. (which I have included in the example project) Where you use the mesh instance material override such that each instance has a copy of the shader and uses that copy to pass in parameters. This is a problem for a couple reasons. 1. I switched from mesh instance to multimesh instance where multimesh instance doesn't have a surface material override yet so it simply isnt available. 2. Making a copy of the whole shader seems a bit extra when all that has changed between instances are the parameters. There has got to be a more effecient solution. 3. If I am applying this material to every instance then it defeats the purpose of using the mesh instance over ride which should only be used if the material being used is the exception rather than the norm.
[local_to_scene_bug.zip](https://github.com/godotengine/godot-proposals/files/6544191/local_to_scene_bug.zip)

### Is there a reason why this should be core and not an add-on in the asset library?
Would probably save resources by not copying shaders.

**Comment by KoBeWi** [comment]
Parameters are saved in the material, not shader.

**Comment by lukostello** [comment]
the material is a visual shader, perhaps I'm confused is there a checkmark I missed to make the parameters local?

**Comment by lukostello** [comment]
looks like I had the material resource checked as local to scene but not the shader but after checking that as local to scene, it doesn't seem to make a difference.

**Comment by KoBeWi** [comment]
I checked your project and the problem is that you use a material saved to a file. It can't be local to scene this way. Right-click it and make unique. (although it doesn't seem to work perfectly either)

In any case, this is something for a bug report rather than a proposal.

**Comment by Xrayez** [comment]
![image](https://user-images.githubusercontent.com/17108460/119648875-a325cf00-be2a-11eb-8606-16e4737b12e3.png)

Looks like workflow issue. Basically when you duplicate the instantiated scene with <kbd>Ctrl + D</kbd>, it will *share* the material. But when you manually *instance* the scene, then it will make the resource unique.

By the way, making a shader local to scene will force duplication of the shader in each instance, which is not desired for performance reasons.

**Comment by lukostello** [comment]
@Xrayez  when I made it, I was aware of the ctrl+d issue and I dragged each instance from the bottom to specifically avoid that. Didn't seem to work. Dragged from the file system a 2nd time to check if I did it wrong the first time and it didn't work again. Is it supposed to have unique materials when dragged from the file system is that something worth making a bug report for?

@KoBeWi  I'm still getting the hang of this make unique thing. Its very hard especially because it looks the same in the inspector before and after you hit make unique so I can't tell if something is already unique in the first place. But I made each material unique in the ones that weren't working before and now they seem to work so thank you very much. I thought having the box checked for "local to scene" here 
![image](https://user-images.githubusercontent.com/53994293/119668067-d5243a80-be04-11eb-82bf-e86d1a310936.png)
 would have made the materials unique in each instant. Perhaps thats for unique aabbs and lightmap size? Its hard to tell which properties the "local to scene" checkmark applies to. Seems like it should make the mesh and default surface material unique in each instance since its in the same box. I'd offer constructive feedback on how the UI could better communicate which properties "local to scene" has an effect on and how to show something is unique as well but I'm fairly certain I don't completely understand it as evidenced by my failure to do this right first try. 

I didn't put it as a bug because there was a strong possibility it was a user error or even if I can't do it I should never be able to do it simply by limitations of coding. In either case it is doing exactly what it is intended to do, so this seemed more like an enhancement to me. I'll file a bug report if you elaborate on what didn't work after you made the materials unique. Seems to have done the trick from my end.

**Comment by Xrayez** [comment]
> Its hard to tell which properties the "local to scene" checkmark applies to. Seems like it should make the mesh and default surface material unique in each instance since its in the same box.

I think it's a general workflow issue which is related to #2737. But once you're aware of Godot's peculiarities, it does make sense, so to speak. 😅

But I do think that "Local To Scene" should only be applicable for a single resource.

See also godotengine/godot#45350.

**Comment by lukostello** [comment]
I just tried deleting the materials to control for the material being saved to a file being related to dragging the instances from the file system causing the materials to not be unique. Not only did it not solve the problem but it made the instance_base which use to work no longer work which was really curious. I don't know what to make of this behavior. I've got no idea if this is a bug or not. I can still hit "make_unique" on each ones material to fix it though. I think part of why I wasn't able to find the "make_unique" solution while devloping is because in my game I am making the scenes on the fly with code and I've go no idea how to make each material unique doing it that way. There doesn't seem to be a method to call to do that.

**Comment by KoBeWi** [comment]
> I've go no idea how to make each material unique doing it that way. There doesn't seem to be a method to call to do that.

Something like `material = material.duplicate()`

**Comment by lukostello** [comment]
@Zireael07 This one? 
![image](https://user-images.githubusercontent.com/53994293/119743717-7e485080-be58-11eb-8612-47b1ca32474f.png)
Doesn't that copy the whole shader code? Seems unnecessary I'll have to test it.

**Comment by lukostello** [comment]
[local_to_scene_bug.zip](https://github.com/godotengine/godot-proposals/files/6551598/local_to_scene_bug.zip)
After having spent the day understanding the nuances between make unique and make local. I am confident this is not intended behavior and is indeed a bug. While making each material unique is a solution. It is far more difficult of a solution than necessarity. It is my belief that if the material's "Local to scene" function worked accurately then every time the "default_base" scene is instanced each instance would have a unique copy of the material and wouldn't need to be made unique manually. 

My hypothesis is that after we check the "local to scene" function of the mesh resource which we want to make sure this box is checked so when we have  a copy of the material it has a unique mesh to apply it to, what is happening is it is first making a copy of the mesh resource which has a reference to the original material in it. Then after that it is making a unique copy of the material. Only it is not being referenced by the mesh resource and because it then doesn't have any users, it is likely being deleted shortly after. 

If my hypothesis is correct, the solution should be simple: make the copy of the material first then make the copy of the mesh resource reference it. I've uploaded a slightly different variation which perhaps highlights this better and eliminates saved materials from being a factor.

**Comment by aaronfranke** [comment]
>@KoBeWi: I checked your project and the problem is that you use a material saved to a file. 

In the latest project @lukostello posted (in the comment above this), the material isn't saved to a file. All of the resources are sub-resources of the scene, and the bug still happens.

I did some testing, and the below diff for 3.x fixes the issue (but probably causes other issues). There is some bogus logic going on causing "local to scene" resources as children of "local to scene" resources to not be duplicated. Without this test code, for the material resource, the commented-out `continue` line is ran. When that is commented out, it's in the cache so it doesn't duplicate. I had to comment out both of these things to hack it into working.

```diff
Ref<Resource> Resource::duplicate_for_local_scene(Node *p_for_scene, Map<Ref<Resource>, Ref<Resource>> &remap_cache) {
	List<PropertyInfo> plist;
	get_property_list(&plist);

	Ref<Resource> r = Object::cast_to<Resource>(ClassDB::instance(get_class()));
	ERR_FAIL_COND_V(r.is_null(), Ref<Resource>());

	r->local_scene = p_for_scene;

	for (List<PropertyInfo>::Element *E = plist.front(); E; E = E->next()) {
		if (!(E->get().usage & PROPERTY_USAGE_STORAGE)) {
-			continue;
		}
		Variant p = get(E->get().name);
		if (p.get_type() == Variant::OBJECT) {
			RES sr = p;
			if (sr.is_valid()) {
				if (sr->is_local_to_scene()) {
-					if (remap_cache.has(sr)) {
-						p = remap_cache[sr];
-					} else {
						RES dupe = sr->duplicate_for_local_scene(p_for_scene, remap_cache);
						p = dupe;
						remap_cache[sr] = dupe;
-					}
				}
			}
		}

		r->set(E->get().name, p);
	}

	return r;
}
```

**Comment by Zireael07** [comment]
@lukostello: Yes, that one. Yes, unfortunately it duplicates the shader code, but I found no other way to get consistent independent materials, both in the racer project and in my other Godot project.

**Comment by Calinou** [comment]
> Yes, that one. Yes, unfortunately it duplicates the shader code, but I found no other way to get consistent independent materials, both in the racer project and in my other Godot project.

[Per-instance uniforms are required for that](https://godotengine.org/article/godot-40-gets-global-and-instance-shader-uniforms), and they're only available in Godot 4.0.

**Comment by lukostello** [comment]
> @lukostello: Yes, that one. Yes, unfortunately it duplicates the shader code, but I found no other way to get consistent independent materials, both in the racer project and in my other Godot project.

![image](https://user-images.githubusercontent.com/53994293/119842557-c73dea80-bed4-11eb-81ed-559a7f7d8e40.png)
I actually learned it doesn't duplicate the shader code, the local to scene that is higher in this picture is the one that duplicates the shader code. The one that is lower and checked duplicates the material which is how the shader is applied to a specific instance, to which end the shader parameters are relevant. Which is why the shader params are outside the shader box and inside the material box as you can see. 

But idk why you would need that check if you are also using material.duplicate() as having the material checked as "local to scene" should make it duplicate itself when a scene with that material is instanced. You should either need to material.duplicate() or "local_to_scene" but not both. Thats why I'm confident this is a bug.

**Comment by lukostello** [comment]
> > Yes, that one. Yes, unfortunately it duplicates the shader code, but I found no other way to get consistent independent materials, both in the racer project and in my other Godot project.
> 
> [Per-instance uniforms are required for that](https://godotengine.org/article/godot-40-gets-global-and-instance-shader-uniforms), and they're only available in Godot 4.0. They can't be backported to `3.x` due to their reliance on backwards-incompatible changes.

If I understand this change correctly this change would no longer you to duplicate materials or make them unique to pass in unique parameters per instance?

**Comment by Calinou** [comment]
> If I understand this change correctly this change would no longer you to duplicate materials or make them unique to pass in unique parameters per instance?

Indeed :slightly_smiling_face:

**Comment by KoBeWi** [comment]
I managed to reproduce the bug and was about to create an issue, but then I checked on master and seems like it's fixed in 4.0. If both mesh and material are local to scene, they are properly made unique upon instancing (which is not the case in 3.x)

**Comment by lukostello** [comment]
dang, can't wait for 4.0
